kyma-incubator/service-catalog
# Add webhook skeleton, remove api-server from chart, add webhoook srv

This PR is a 

**What this PR does / why we need it**:

- Add webhook skeleton
- Remove api-server from chart
- Add webhook server in chart
- Move PrepareForCreate login into webhook handlers
- Move logic from defaults into webhook handlers

**Which issue(s) this PR fixes** 

- https://github.com/kyma-project/kyma/issues/2784


#### NOTE
Right now u can see that the image was pushed to the kyma-project repository. This is not automated, here are the commands used to do that:
```
make service-catalog-image
docker tag service-catalog-amd64:canary eu.gcr.io/kyma-project/develop/service-catalog/service-catalog-amd64:crd-0.0.1
docker push eu.gcr.io/kyma-project/develop/service-catalog/service-catalog-amd64:crd-0.0.1
```

How was tested:
- Bootstrap environment using [this](https://github.com/kyma-incubator/service-catalog/tree/crds-on-v0.1.40/contrib/hack/crd/bin) script.
- Install service catalog
  ```
  helm install --name catalog --namespace kyma-system charts/catalog/
  ```
- Execute [this](https://github.com/kyma-incubator/service-catalog/tree/crds-on-v0.1.40/contrib/hack/crd#testing-scenario) example scenario. 
  NOTE: 
  - `assets/scenario/redis-instance-manual.yaml` change to `assets/scenario/redis-instance.yaml` 
  - `assets/scenario/redis-instance-binding-manual.yaml` change to `assets/scenario/redis-instance-binding.yaml`



I also executed such comparison:
I prepared [sample CRs](https://gist.github.com/mszostok/590fbe64961c8d754df6920021c5e73c)

Then on old service catalog installation, I scaled to 0 the controller-manager and left only the api-server. Next, I created all prepared sample CRs, so then I got all objects in etcd which were initialized by api-server.

The same scenario I executed for new Service Catalog version. I applied CRDs and deployed webhook server and then created all prepared sample CRs, so then I got all objects in etcd which were initialize by webhooks mutations.

I expected to have the same end result.


| Type  | Webhook  |  API-Server | 
|---|---|---|
| ClusterServiceBroker  | no `Status` entry  | initialized `Status` entry with zero-values |
| ClusterServiceClass  |  contains fields only from sample CR | contains fields from Sample obj + rest fields with zero values + missing **generation** field(bug) |
|  ClusterServicePlan |  contains fields only from sample CR |  contains fields from Sample obj + rest fields with zero values + missing **generation** field(bug) |
|  ServiceInstance |  contains fields only from sample CR, no `Status` entry |  contains fields from Sample obj + rest fields with zero values + initialized `Status` entry |
|  ServiceBinding |  contains fields only from sample CR, no `Status` entry |  contains fields from Sample obj + initialized `Status` entry |
|  ServiceBroker |  contains fields only from sample CR, no `Status` entry |  contains fields from Sample obj + initialized `Status` entry with zero-values |
| ServiceClass  |  contains fields only from sample CR | contains fields from Sample obj + rest fields with zero values  |
| ServicePlan |  contains fields only from sample CR |  contains fields from Sample obj + rest fields with zero values |

#### Found diff:
- `Status` entry cannot be initialized in webhook on CREATE action, because of that this logic will be moved to controller (will be in separate PR)
- ClusterServiceClass/ClusterServicePlan and ServiceClass/ServicePlan have fields with zero value. This can me mitigated because in normal flow those CRs are created by controller so spec will the same as previously. Changes are visible only if they are created manually by user - nto applicable in normal flow)
